File: sample-rxjava-java/src/main/java/ru/solodovnikov/rxlocationmanager/sample/MainActivity.java
package ru.solodovnikov.rxlocationmanager.sample;

import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import java.util.concurrent.TimeUnit;

import ru.solodovnikov.rxlocationmanager.IgnoreErrorTransformer;
import ru.solodovnikov.rxlocationmanager.LocationRequestBuilder;
import ru.solodovnikov.rxlocationmanager.LocationTime;
import ru.solodovnikov.rxlocationmanager.RxLocationManager;
import rx.Single;
import rx.functions.Action1;

public class MainActivity extends AppCompatActivity {
    private RxLocationManager rxLocationManager;
    private LocationRequestBuilder locationRequestBuilder;

    private CoordinatorLayout coordinatorLayout;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        rxLocationManager = new RxLocationManager(this);
        locationRequestBuilder = new LocationRequestBuilder(rxLocationManager);

        coordinatorLayout = (CoordinatorLayout) findViewById(R.id.root);

        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.last_network:
                requestLastNetworkLocation();
                return true;
            case R.id.last_network_minute_old:
                requestLastNetworkOneMinuteOldLocation();
                return true;
            case R.id.request_location:
                requestLocation();
                return true;
            case R.id.complicated_request_location:
                requestBuild();
                return true;
            case R.id.complicated_request_location_ignore_error:
                requestBuildIgnoreSecurityError();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void requestLastNetworkLocation() {
        testSubscribe(rxLocationManager.getLastLocation(LocationManager.NETWORK_PROVIDER), "requestLastNetworkLocation");
    }

    private void requestLastNetworkOneMinuteOldLocation() {
        testSubscribe(rxLocationManager.getLastLocation(LocationManager.NETWORK_PROVIDER, new LocationTime(1, TimeUnit.MINUTES)), "requestLastNetworkOneMinuteOldLocation");
    }

    private void requestLocation() {
        testSubscribe(rxLocationManager.requestLocation(LocationManager.NETWORK_PROVIDER, new LocationTime(15, TimeUnit.SECONDS)), "requestLocation");
    }

    private void requestBuild() {
        final Single<Location> single = locationRequestBuilder
                .addLastLocation(LocationManager.NETWORK_PROVIDER, new LocationTime(30, TimeUnit.MINUTES))
                .addRequestLocation(LocationManager.NETWORK_PROVIDER, new LocationTime(15, TimeUnit.SECONDS))
                .setDefaultLocation(new Location(LocationManager.PASSIVE_PROVIDER))
                .create();

        testSubscribe(single, "requestBuild");
    }

    private void requestBuildIgnoreSecurityError() {
        final IgnoreErrorTransformer ignoreErrorTransformer = new IgnoreErrorTransformer(SecurityException.class);

        final Single<Location> single = locationRequestBuilder
                .addLastLocation(LocationManager.NETWORK_PROVIDER, new LocationTime(30, TimeUnit.MINUTES), ignoreErrorTransformer)
                .addRequestLocation(LocationManager.NETWORK_PROVIDER, new LocationTime(15, TimeUnit.SECONDS), ignoreErrorTransformer)
                .setDefaultLocation(new Location(LocationManager.PASSIVE_PROVIDER))
                .create();

        testSubscribe(single, "requestBuild");
    }

    private void showSnackbar(CharSequence text) {
        Snackbar.make(coordinatorLayout, text, Snackbar.LENGTH_SHORT)
                .show();
    }

    private void testSubscribe(Single<Location> single, final String methodName) {
        single.subscribe(new Action1<Location>() {
            @Override
            public void call(Location location) {
                showLocationMessage(location, methodName);
            }
        }, new Action1<Throwable>() {
            @Override
            public void call(Throwable throwable) {
                showErrorMessage(throwable, methodName);
            }
        });
    }

    private void showLocationMessage(Location location, String methodName) {
        final String pattern = "%s Success: %s";
        showSnackbar(String.format(pattern, methodName, location != null ? location.toString() : "Empty location"));
    }

    private void showErrorMessage(Throwable throwable, String methodName) {
        final String pattern = "%s Error: %s";
        showSnackbar(String.format(pattern, methodName, throwable.getMessage()));
    }
}

File: sample-rxjava2-kotlin/src/main/kotlin/ru/solodovnikov/rxlocationmanager/sample/MainActivity.kt
package ru.solodovnikov.rxlocationmanager.sample

import android.content.Intent
import android.location.Location
import android.location.LocationManager
import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.Snackbar
import android.support.v4.app.ActivityCompat
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import io.reactivex.Maybe
import io.reactivex.Single
import ru.solodovnikov.rx2locationmanager.*
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity(), BasePermissionBehavior.PermissionCallback, ForResultCaller {
    private val rxLocationManager: RxLocationManager by lazy { RxLocationManager(this) }
    private val locationRequestBuilder: LocationRequestBuilder by lazy { LocationRequestBuilder(rxLocationManager) }

    private val coordinatorLayout by lazy { findViewById<CoordinatorLayout>(R.id.root) }

    private var checkPermissions = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(findViewById(R.id.toolbar))
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.check_permissions -> {
                item.isChecked = !item.isChecked
                checkPermissions = item.isChecked
                return true
            }
            R.id.last_network -> {
                requestLastNetworkLocation()
                return true
            }
            R.id.last_network_minute_old -> {
                requestLastNetworkOneMinuteOldLocation()
                return true
            }
            R.id.request_location -> {
                requestLocation()
                return true
            }
            R.id.complicated_request_location -> {
                requestBuild()
                return true
            }
            R.id.complicated_request_location_ignore_error -> {
                requestBuildIgnoreSecurityError()
                return true
            }
            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_LOCATION_SETTINGS) {
            rxLocationManager.onActivityResult(resultCode, data)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_LOCATION_PERMISSIONS) {
            rxLocationManager.onRequestPermissionsResult(permissions, grantResults)
        }
    }

    override fun requestPermissions(permissions: Array<String>) {
        ActivityCompat.requestPermissions(this, permissions, REQUEST_CODE_LOCATION_PERMISSIONS)
    }

    private fun requestLastNetworkLocation() {
        if (checkPermissions) {
            rxLocationManager.getLastLocation(LocationManager.NETWORK_PROVIDER, behaviors = *arrayOf(EnableLocationBehavior.create(this, REQUEST_CODE_LOCATION_SETTINGS, { this }, rxLocationManager), PermissionBehavior(this, rxLocationManager, this)))
        } else {
            rxLocationManager.getLastLocation(LocationManager.NETWORK_PROVIDER)
        }.testSubscribe("requestLastNetworkLocation")
    }

    private fun requestLastNetworkOneMinuteOldLocation() {
        if (checkPermissions) {
            rxLocationManager.getLastLocation(LocationManager.NETWORK_PROVIDER, LocationTime(1, TimeUnit.MINUTES), PermissionBehavior(this, rxLocationManager, this))
        } else {
            rxLocationManager.getLastLocation(LocationManager.NETWORK_PROVIDER, LocationTime(1, TimeUnit.MINUTES))
        }.testSubscribe("requestLastNetworkOneMinuteOldLocation")
    }

    private fun requestLocation() {
        if (checkPermissions) {
            rxLocationManager.requestLocation(LocationManager.NETWORK_PROVIDER, LocationTime(15, TimeUnit.SECONDS), PermissionBehavior(this, rxLocationManager, this))
        } else {
            rxLocationManager.requestLocation(LocationManager.NETWORK_PROVIDER, LocationTime(15, TimeUnit.SECONDS))
        }.testSubscribe("requestLocation")
    }

    private fun requestBuild() {
        if (checkPermissions) {
            val permissionTransformer = PermissionBehavior(this, rxLocationManager, this)
            locationRequestBuilder
                    .addLastLocation(LocationManager.NETWORK_PROVIDER, LocationTime(30, TimeUnit.MINUTES), permissionTransformer)
                    .addRequestLocation(LocationManager.NETWORK_PROVIDER, LocationTime(15, TimeUnit.SECONDS), permissionTransformer)
                    .setDefaultLocation(Location(LocationManager.PASSIVE_PROVIDER))
                    .create()
        } else {
            locationRequestBuilder
                    .addLastLocation(LocationManager.NETWORK_PROVIDER, LocationTime(30, TimeUnit.MINUTES))
                    .addRequestLocation(LocationManager.NETWORK_PROVIDER, LocationTime(15, TimeUnit.SECONDS))
                    .setDefaultLocation(Location(LocationManager.PASSIVE_PROVIDER))
                    .create()
        }.testSubscribe("requestBuild")
    }

    private fun requestBuildIgnoreSecurityError() {
        val ignoreError = IgnoreErrorBehavior(SecurityException::class.java)

        locationRequestBuilder
                .addLastLocation(LocationManager.NETWORK_PROVIDER, LocationTime(30, TimeUnit.MINUTES), ignoreError)
                .addRequestLocation(LocationManager.NETWORK_PROVIDER, LocationTime(15, TimeUnit.SECONDS), ignoreError)
                .setDefaultLocation(Location(LocationManager.PASSIVE_PROVIDER))
                .create()
                .testSubscribe("requestBuild")
    }

    private fun showSnackbar(text: CharSequence) {
        Snackbar.make(coordinatorLayout, text, Snackbar.LENGTH_SHORT)
                .show()
    }

    fun Maybe<Location>.testSubscribe(methodName: String) {
        subscribe({ showLocationMessage(it, methodName) },
                { showErrorMessage(it, methodName) },
                { showSnackbar("$methodName Completed") })
    }

    fun Single<Location>.testSubscribe(methodName: String) {
        subscribe({ showLocationMessage(it, methodName) },
                { showErrorMessage(it, methodName) })
    }

    private fun showLocationMessage(location: Location?, methodName: String) {
        showSnackbar("$methodName Success: ${location?.toString() ?: "Empty location"}")
    }

    private fun showErrorMessage(throwable: Throwable, methodName: String) {
        showSnackbar("$methodName Error: ${throwable.message}")
    }

    companion object {
        private const val REQUEST_CODE_LOCATION_PERMISSIONS = 150
        private const val REQUEST_CODE_LOCATION_SETTINGS = 151
    }
}
